src/cargo/ops/cargo_new.rs

```diff
@@ -67,22 +64,144 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
                                   c, name)));
     }
+    Ok(())
+}
+
+fn detect_source_path_and_type<'a : 'b, 'b>(project_path : &Path, 
+                                            project_name: &str, 
+                                            opts2: &'b mut NewOptions) -> CargoResult<()> {
+    let path = project_path;
+    let name = project_name;
+    
+    let mut found_source_files = 0;
+    
+    if paths::file_already_exists(&path.join("src/main.rs")) {
+        opts2.bin = true;
+        opts2.sourcefile_relative_path = Some(String::from("src/main.rs"));
+        found_source_files += 1;
+    }
+    if paths::file_already_exists(&path.join("main.rs")) {
+        opts2.bin = true;
+        opts2.sourcefile_relative_path = Some(String::from("main.rs"));
+        found_source_files += 1;
+    }
+    fn autodetect_bin_file(p: &Path) -> CargoResult<bool> {
+        let mut content = String::new();
+        try!(File::open(p).and_then(|mut x| x.read_to_string(&mut content)));
+        Ok(content.contains("fn main"))
+    }
+    if paths::file_already_exists(&path.join(format!("{}.rs", name))) {
+        if opts2.bin {
+            // OK
+        } else {
+            opts2.bin = try!(autodetect_bin_file(&path.join(format!("{}.rs", name))));
+            
+        }
+        opts2.sourcefile_relative_path = Some(format!("{}.rs", name));
+        found_source_files += 1;
+    }
+    
+    if paths::file_already_exists(&path.join(format!("src/{}.rs", name))) {
+        if opts2.bin {
+            // OK
+        } else {
+            opts2.bin = try!(autodetect_bin_file(&path.join(format!("src/{}.rs", name))));
+            
+        }
+        opts2.sourcefile_relative_path = Some(format!("src/{}.rs", name));
+        found_source_files += 1;
+    }
+    
+    if (!opts2.bin) && paths::file_already_exists(&path.join("src/lib.rs")) {
+        found_source_files += 1;
+        opts2.sourcefile_relative_path = Some(String::from("src/lib.rs"));
+    }
+    if (!opts2.bin) && paths::file_already_exists(&path.join("lib.rs")) {
+        found_source_files += 1;
+        opts2.sourcefile_relative_path = Some(String::from("lib.rs"));
+    }
+    
+    if found_source_files > 1 {
+        return Err(human(r"There are multiple eligible source files for `cargo init`.
+I don't know which Cargo.toml template to use."));
+    }
+    Ok(())
+}
+
+pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
+    let path = config.cwd().join(opts.path);
+    if fs::metadata(&path).is_ok() {
+        return Err(human(format!("Destination `{}` already exists",
+                                 path.display())))
+    }
+    
+    let name = try!(get_name(&path, &opts, config));
+    try!(check_name(name));
+    
     mk(config, &path, name, &opts).chain_error(|| {
         human(format!("Failed to create project `{}` at `{}`",
                       name, path.display()))
     })
 }
 
+pub fn init(opts: NewOptions, config: &Config) -> CargoResult<()> {
```

Comment:
Currently other commands all take the struct of options via a reference, so perhaps that could happen here as well?